Grid filtering refactor for improved accuracy and persistence via URL parameters

smg6511 opened the pull request:
### What does it do?
Built upon work merged in #16089 and #16355 to provide better filtering functionality and persistence for the majority of grids. Additional methods were added to the Tabs and Grids base classes to further optimize the code (_e.g._, centralizing the repetitive task of creating the query and clear buttons, which are for the most part exactly the same everywhere they are applied). 

### Why is it needed?
To apply the new global persistence feature to the referenced grids and improve their UIs.

### Grids covered

1. **Media Sources** - Media > Media Sources (/manager/?a=source)
2. **Packages** - Extras > Installer (/manager/?a=workspaces)
3. **Users** - Manage > Users (/manager/?a=security/user)
4. **User Settings** - Manage > Users > [User] Settings (manager/?a=security/user/update&id=1)
5. **Messages** - User (user icon) > Messages (/manager/?a=security/message)
6. **System Settings** - Settings (gear) > System Settings (/manager/?a=system/settings)
7. **System Events** - Settings (gear) > System Settings, 2nd tab (/manager/?a=system/settings)
8. **Form Customization Profiles** - Settings (gear) > Form Customization (/manager/?a=security/forms)
9. **Profile** - Settings (gear) > Form Customization > [Profile] (manager/?a=security/forms/profile/update&id=1)
10. **Dashboards** - Settings (gear) > Dashboards (/manager/?a=system/dashboards)
11. **Widgets** - Settings (gear) > Dashboards, 2nd tab (/manager/?a=system/dashboards)
12. **Contexts** - Settings (gear) > Contexts (/manager/?a=context)
13. **Context Settings** - Settings (gear) > Contexts > [Context], 2nd tab (/manager/?a=context/update&key=[yourcontext])
14. **User Groups & Users** - Settings (gear) > Access Control Lists [click UG in left tree] (/manager/?a=security/permission&group=1)
15. **Access Policies** - Settings (gear) > Access Control Lists, 3rd tab (/manager/?a=security/permission)
16. **Policy Templates** - Settings (gear) > Access Control Lists, 4th tab (/manager/?a=security/permission)
17. **Access Permissions/Contexts** - Settings (gear) > Access Control Lists > [User Group], 2nd tab (manager/?a=security/usergroup/update&id=1)
18. **Access Permissions/Resource Groups** - (same path as # 17, 2nd vertical tab)
19. **Access Permissions/Element Categories** - (same path as # 17, 3rd vertical tab)
20. **Access Permissions/Media Sources** - (same path as # 17, 4th vertical tab)
21. **Access Permissions/Namespaces** - (same path as # 17, 5th vertical tab)
22. **User Group/Users** - Settings (gear) > Access Control Lists > [User Group], 3rd tab (manager/?a=security/usergroup/update&id=1)
23. **User Group/Settings** - Settings (gear) > Access Control Lists > [User Group], 4th tab (manager/?a=security/usergroup/update&id=1)
24. **Lexicon Management** - Settings (gear) > Lexicons (/manager/?a=workspaces/lexicon)
25. **Namespaces** - Settings (gear) > Namespaces (/manager/?a=workspaces/namespace)
26. **Trash Bin** - [Trash icon in Resources/Elements/Categories tab panel] (/manager/?a=resource/trash)
27. **Template/Template Variables** - Elements Tab > [Template], 2nd tab (/manager/?a=element/template/update&id=1)
28. **Template Variables/Template Access** - Elements Tab > [Template Variable], 4th tab (/manager/?a=element/tv/update&id=1)
29. **Plugins/System Events** - Elements Tab > [Plugin], 2nd tab (/manager/?a=element/plugin/update&id=1)

### How to test

1. Ensure all caches a cleared.
2. Visit all grids with filtering capabilities and perform various search queries and clear them, verifying expected behavior.
3. For each grid, perform a search and copy the URL in your browser. Paste that URL in a new window or browser tab to verify navigation to the correct tab, and that the search query is intact.

### Related issue(s)/PR(s)
Replaces the closed PRs: #16081, #16084, #16063, #15946, #15942, and #15935

Commit:
Progress round 2, manual conflict resolution on base grid and tabs classes

Commit:
Progress round 2, manual conflict resolution for usergroup acls panels

Commit:
Revert "Progress round 2, manual conflict resolution on base grid and tabs classes"

This reverts commit d70885cc3ecfcf97f80d39300aad2df84fcf1fdd.

Commit:
Base class updates required for grid enhancements

Contains new required methods and tweaks for grid updates to follow

Commit:
Filter Persistence: Policies and Policy Templates

Mark-H commented on manager/assets/modext/core/modx.js:
```diff
@@ -551,7 +551,6 @@ Ext.extend(MODx,Ext.Component,{
             });
             tabPanel.add(newTabConfig);
             tabPanel.doLayout();
-            tabPanel.setActiveTab(0);
```
Ohhhhh this must be why VersionX always kept forcing the first tab to open, breaking the functionality that keeps tabs open. Never figured that out, but that does use this method 😂

smg6511 replied:
Yes, I was kinda pulling my hair out for a while there because last I worked on this all was well. But in between I did install VersionX and after digging found that this line was causing havoc. I poked around to see if it was really needed, and found it was not.

Mark-H commented on manager/assets/modext/util/utilities.js:
```diff
@@ -548,12 +548,12 @@ MODx.util.url = {
      * @param {Object} stateData - Optional data to be used in subsequent url processing
      */
-    clearParams: function(stateData = {}) {
+    clearAllParams: function(stateData = {}) {
```
Changes like this can be considered a breaking change, because it's a public APIs that's suddenly a different name. However this is one that was added in your previous PR, right? So that hasn't been released yet?

smg6511 replied:
Correct, that's new and would not have been used anywhere else, yet.

Mark-H commented on manager/assets/modext/widgets/core/modx.tabs.js:
```diff
@@ -1,41 +1,56 @@
+                the state of the Resource tab panel can become uncertain and no tab will be initially
+                selected. This workaround ensures the first tab is selected.
+            */
+            if (this.id === 'modx-resource-tabs' && MODx.request.tab === undefined && !this.getActiveTab()) {
```
If we want to try and compose things properly, this sort of edge-case overrides should be in the component that experiences the edge-case. I.e. in the `modx-resource-tabs` component, not here in `MODx.Tabs`. 

However, all of the ExtJS stuff is already messy and not necessarily composed properly in many places, so I'm _very_ low on the SSOGAF about this. If you think this is the easiest way to go about this particular issue, then that's fine and you can mark this resolved without making changes.

smg6511 replied:
I'll take a look and let you know later today.

smg6511 replied:
Good catch here @Mark-H. Indeed this is better placed in the Resources panel. Done in [this](https://github.com/modxcms/revolution/pull/16369/commits/559f676a2bd24ef938878c033a73907d12ed554c) commit.

Mark-H commented:
> Many other grids have these enhancements ready to submit, but rely on the base class changes committed in this PR. As soon as this is tested and merged, I'll put up additional PRs to cover other grids.

If the changes to the other grids are the same as the updated grids already here (i.e. like 20 lines of changes (ignoring whitespace) that only re-define the tbar), feel free to add those onto this PR so it can all be tested at once. I've developed quite the distaste for this concept of dependent PRs and it's not thousands of lines.

smg6511 commented:
@Mark-H - Let me ask you this: I have several other changed grids updated to use this feature (remember that this global handling will ultimately apply to all grids that have filtering). In all, excluding the files changed in this PR, I have ~45 changed files (involving roughly 13 grids) in my local work in progress; most of them don't involve big changes, but it's still a lot of grids to test. How many would you want submitted in this PR? I'm happy to add as much as you think can practically be tested at once.

opengeek commented:
I agree with @Mark-H on including all of these changes related to the "base class changes" introduced here in one PR. Having PRs withheld waiting on a base change, unless the base change is being submitted separately and can be evaluated in isolation from changes expected to be made on top of it, creates more complication than reviewing a larger-sized PR.

Mark-H commented:
There is no one-size-fits all. There is no number of lines or number of files that makes one PR harder to review. 

There is a mental load to keep in mind. 13 small PRs that are all essentially the same except it affects a different file, are IMO harder to review and test than one PR that changes all the grids, not in the least because you're then in need of 26 approvals vs just 2, but also because every PR requires rebuilding the mental model of what is happening. Context switching is expensive, so as long as it's all the same context, combining makes sense. 

Testing also. I'm definitely not going to spin up a test environment 13 times for the individual changes because the time to set that up will exceed the time it takes to test it. But I will spin up the test environment for a single bigger PR and take the time to test each grid. 

It becomes a different story when we're starting to make a wide range of changes with different goals. Then it makes sense to break it up so one pr = one goal. 

So, as long as your per-grid changes (e.g. manager/assets/modext/widgets/security/modx.grid.access.policy.js) are all very similar and do not involve individual refactors (excluding whitespace) exceeding a couple dozen lines, just dump 'm all in, we'll review all of it once, and get this whole multi-month saga over with already.

smg6511 commented:
@opengeek @Mark-H - Ok, I'm totally down with that. I think I still question how to go about this because of mixed signals earlier on (not necessarily from you guys) with a couple of my larger PRs. I'll spin up the rest of what I've got ready and add it in here. Is it at all useful to you for me to make multiple commits organized by grid (having 2 or 3 files at the most as tweaks to each grid's processor were made)?

Mark-H commented:
Yes, atomic commits (where one commit = one finished change for one specific item) are always a good idea. In the end it'll all get squashed per PR but if there is additional context to be had, great.

Commit:
Filter Persistence: Deleted Resources Manager (Trash)

Also applied refinement to combo filter (showing only relevant Contexts) and fixed query so only deleted Resources are returned (fix from Ruslan's closed PR).

Commit:
Update user group permissions, category grid

Applies new shared method to create the clear filters button

Commit:
Update user group permissions, contexts grid

Applies new shared method to create the clear filters button

Commit:
Update user group permissions, namespace grid

Applies new shared method to create the clear filters button

Commit:
Update user group permissions, resource groups grid

Applies new shared method to create the clear filters button

Commit:
Update user group permissions, media sources grid

Applies new shared method to create the clear filters button

Commit:
Filter Persistence: Plugin Events Grid

Commit:
Move resource panel-specific tabchange listener

Since the fix is only relevant to one panel, moved it out of the Tabs base class

Commit:
Filter Persistence: Namespaces Grid

Commit:
Filter Persistence: Contexts Grid

Commit:
Filter Persistence: Form Customization Profiles Grid

Commit:
Filter Persistence: Form Customization Sets Grid

Commit:
Filter Persistence: Dashboards Grid

Commit:
Filter Persistence: Dashboard Widgets Grid

Commit:
Filter Persistence: Packages Grid

Commit:
Filter Persistence: Media Sources Grid

Commit:
Update related to commit 0e0778d

Forgot to add this file in for Contexts panel/grid

Commit:
Update related to commit c7fd92b

Forgot to add this file

Commit:
Filter Persistence: Users Grid

Commit:
Filter Persistence: Groups Panel and Users Grid

Note that this is the panel that appears upon first going to the ACLs. The affected grid appears when clicking on a User Group in the left tree. The grid also appears under the Users panel when editing a User Group.

Commit:
Filter Persistence: Template TVs Grid

Also fixes processor so the grid total count is always correct, making the paging correct and navigable.

Commit:
Filter Persistence: TV Templates Grid

JoshuaLuckers commented:
@opengeek @Mark-H I think the release with these changes should be released as a beta first.

Mark-H commented:
Yeah, always in favor of at least one release candidate for minor releases anyway. :) 

I've been keeping track of the commits added so far and am not seeing any issues yet.

smg6511 commented:
@Mark-H - Couple questions:

1. Roughly how long will it be before we get to a beta/rc for 3.1? I ask because I want to be sure to get in the last two major grids beforehand (lexicons and system settings); those are the most complicated of all of them and I have a little more tweaking to do there before committing them.
2. If I perform a rebase now will that cause any issues for you? I ask because there are two grids with new filtering capability that just recently got merged (after I began putting this PR together) that I'd like to get in as well. See #16219.

opengeek commented:
We don't do betas for minor releases, and generally don't do RC's for them, either. But if we want to do an RC cycle for minors, that can certainly be accommodated. There are no specific timelines set for this, but I would hope we could get this complete feature set worked out before we release.

You do not need to rebase anything unless there are conflicts or there are JS changes that would affect this functionality. What grids with filtering are you referring to?

smg6511 commented:
@opengeek - I'm referring to the ones merged in #16219. I suppose I could just copy those two classes as they stand and make changes to them in this branch, avoiding the need for a rebase.

Mark-H commented:
Rebasing is a way to deal with conflicts, as-is this branch isn't conflicting yet. However if you start editing the files modified in #16219 again, either making the same or different changes that affect the same lines, then that _will_ cause conflicts and a rebase will become necessary. Does that help decide on how to proceed at this point?

smg6511 commented:
So it seems to me if I _can_ rebase, I should if I want to get those two files in line with this PR (as one way or another it would become necessary). Does that sound reasonable? (BTW: I'm finding that there is something there to be able to rebase.)

Commit:
Filter Persistence: User Messages Grid

mishanthrop commented on manager/assets/modext/util/utilities.js:
```diff
@@ -570,8 +570,62 @@ MODx.util.url = {
+     */
+    clearParam: function(reference, referenceIsComponent = true, stateData = {}) {
+        if (typeof window.history.replaceState !== 'undefined') {
+            const urlParts = window.location.href.split('?'),
+                  removeParamName = referenceIsComponent ? this.getParamNameFromCmp(reference) : reference.trim(),
+                  regex = new RegExp(`${removeParamName}=[^&]+`, 'i')
+            ;
+            let params = urlParts[1].replace(regex, '').replace('&&', '&');
+            if (params.endsWith('&')) {
+                params = params.substr(0, params.length - 1);
+            }
+            let newUrl = new URL(`${urlParts[0]}?${params}`);
+            newUrl = newUrl.toString().replace(/%2F/g, '/');
```
```javascript
const url = new URL(window.location.href)
url.searchParams.delete(removeParamName)
```

smg6511 replied:
Ok, using the URL API does help to simplify it a bit. I updated the method, although it needed more than just the two lines you suggested.

mishanthrop commented on manager/assets/modext/widgets/core/modx.grid.js:
```diff
@@ -907,32 +939,47 @@ Ext.extend(MODx.grid.Grid,Ext.grid.EditorGridPanel,{
-            const id = itemId.trim(),
-                  cmp = this.getFilterComponent(id)
+        data.forEach(item => {
+            itemData = item.replace(/\s+/g, '').split(':');
```
Where is `itemData` variable declaration?

smg6511 replied:
Good catch - fixed that on my side.

Commit:
Filter Persistence: Settings Grids (4)

Includes system settings, usergroup settings, context settings, and user settings. Also adds new method to utilities and makes minor updates for the base grid class.

Commit:
Filter Persistence: System Events Grid

Commit:
Filter Persistence: Lexicons Grid

Also provides two new helper methods in the base modLexicon class to improve filtering granularity, as well as a small styling tweak for the top toolbar.

Commit:
Update UserGroup Namespaces Access Grid

Missed adding this change to commit 504015e

Commit:
Remove Ext Statefulness from TVs Panel

Avoids conflicts with our new tab tracking functionality, and it's generally less confusing to have the tab panel initially show its first tab instead of the last one visited

Commit:
Minor refinements for better Extras suppport

Changes make identification of grid objects within a tab panel more robust/flexible

JoshuaLuckers commented on core/lexicon/en/default.inc.php:
```diff
@@ -314,7 +315,7 @@
 $_lang['orm_container_rename'] = 'Rename Container';
 $_lang['orm_container_remove'] = 'Delete Container';
 $_lang['orm_container_remove_confirm'] = 'Are you sure you want to delete this container and all attributes below it? This is irreversible.';
-$_lang['pagetitle'] = 'Resource\'s title';
+$_lang['pagetitle'] = 'Resource‘s Title';
```
I believe this should stay a `'` and not a `‘`.

smg6511 replied:
Ok, I changed this ... but at some point I'd love to see us find a way to display proper typographical quotes in the interface texts ;-)

JoshuaLuckers commented on core/src/Revolution/Processors/Workspace/PackageNamespace/GetList.php:
```diff
@@ -55,33 +62,94 @@ public function initialize()
+                    $nsSubquery = $this->modx->newQuery($settingsClass);
+                    $nsSubquery->setClassAlias($alias);
+                    $nsSubquery->select([
+                        'namespaces' => "GROUP_CONCAT(DISTINCT `{$alias}`.`namespace` SEPARATOR '\",\"')"
```
Is `SEPARATOR` something special in queries?

smg6511 replied:
Yes, in the CONCAT aggregate functions the `SEPARATOR` is inserted between the values found. The default separator is just a comma, but I need the double quotes as well since the concatenated list being built consists of strings and not integers, and the found list is ultimately being used for the `where IN` condition later on.

Commit:
Fix code quality issues

Commit:
Fix one missed CQ problem

Commit:
Make requested review change

JoshuaLuckers reviewed (commented):
When filtering in Access Policies e.g: `manager?a=security/permission&query=cl&tab=2` the Clear Filter doesn't work.
```
[Error] getFilterComponent: The filter component with itemId 'filter-query' could not be retrieved.

	getFilterComponent (modx.grid.js:1014)
	(anonymous function) (modx.grid.js:957)
	forEach
	clearGridFilters (modx.grid.js:953)
	fn (modx.grid.js:1150)
	fire (ext-all.js:21:3553)
	fireEvent (ext-all.js:21:674)
	onClick (ext-all.js:21:407761)
	A (ext-all.js:21:52936)
```

```
[Error] TypeError: undefined is not an object (evaluating 'cmp.itemId')

	getParamNameFromCmp (utilities.js:607)
	(anonymous function) (modx.grid.js:958)
	forEach
	clearGridFilters (modx.grid.js:953)
	fn (modx.grid.js:1150)
	fire (ext-all.js:21:3553)
	fireEvent (ext-all.js:21:674)
	onClick (ext-all.js:21:407761)
	A (ext-all.js:21:52936)
```

When I filter in `User Group/Users` and go to the settings tab the search filter in settings is "empty" (as it should).
But when I copy the URL after filtering in the Users tab and the open it and go to settings, the search query is also set in the Settings filter. 

So opening the following URL: `manager?a=security/usergroup/update&id=1&query=te&tab=2` will show the filter in the Users tab. But when you switch to the Settings tab it's also in the search field there.
This behavior seems to be across various other grids as well (e.g opening `manager?a=security/usergroup/update&id=1&policy=8&vtab=3&tab=1` and switching to another vtab will also apply the `policy` filter from the previous tab.)

Other than that I didn't encounter any other issues.

smg6511 commented:
@JoshuaLuckers - Ok, great - thanks for giving this monster of a PR a look! I fixed the first issue (which also was occurring with the Policy Templates grid) and am taking a look at the second one.

smg6511 committed:
Update modx.panel.template.js

Remove legacy filtering param

smg6511 committed:
Required base class and utilities updates

New methods and fixes to address issues found in PR review

smg6511 committed:
Settings grid updates

Fixes and tweaks to address PR review issues

smg6511 committed:
ACL tab content updates

Fixes to address PR review issue

smg6511 committed:
Dashboards grid fixes

Fixes to address PR review issues

smg6511 committed:
ACLs - User Group Permissions fixes

Fixes to address PR review issues

smg6511 commented:
@JoshuaLuckers - I finally was able to focus on the remaining issue you'd pointed out in your review and all _should_ be fixed and ready to go. Please take a second look when you get a chance.
@Ruslan-Aleev - Since this PR was motivated by the work you had begun and your request for a global solution, you may want to review this as well if you've got the time ;-)

theboxer reviewed (changes requested):
@smg6511 Impressive work on this one, here are my notes from testing:

### ?a=workspaces
- switching to `Providers` tab doesn't clear the query params
### ?a=security/message
- clear filter causes console error & doesn't clear the search field
### ?a=system/settings
- search field is shared between system settings & system events
### ?a=context/update&key=mgr
- After adding/editing/deleting new context setting, an error shows up in console:
	- The filter component with itemId 'filter-namespace' could not be retrieved
### ?a=security/usergroup/update&id=1
- Access Permissions
	- Contexts
		- Double reload clears the URL params
	- Element Categories
		- Double reload clears the URL params
		- Filter by category doesn't persist correctly (after reload the field is empty)
	- Media Sources
		- Double reload clears the URL params
	- Namespaces
		- Double reload clears the URL params
- Users
	- Double reload clears search field
- Settings
	- After adding/editing/deleting new context setting, an error shows up in console:
		- The filter component with itemId 'filter-namespace' could not be retrieved

smg6511 commented:
@theboxer - Thanks for looking this over closely. I'll sift through the issues you found and provide fixes later this week...

smg6511 committed:
Post-review fixes for system settings grids

smg6511 committed:
Post-review fixes for packages grid

Adds new component id for Package grid that matches the pattern of all others ('modx-grid-[x]')

smg6511 committed:
UG Category tweaks

Minor changes made to the category filter combo to make it correctly display category names when persisted via state/URL

smg6511 committed:
Base class updates

Additional logic to fix issues raised in PR review

smg6511 commented:
@theboxer - Ok, I believe I've resolved all the issues you raised. Please take a look when you get a chance. Also, if you can find the time, it'd be great if you could look at #16469 as well ;-)

theboxer commented:
@smg6511 I hope I'll get to this this week, sorry for the delay

theboxer reviewed (approved):
Seems good now, thanks!

Mark-H commented:
Thank you for your hard work on this @smg6511 - took a while to get it done, but we got there, eventually. Appreciate your _persistence_!

Mark-H commented:
GitHub isn't auto closing because of the way I did the merge (resolving conflicts), but this has been done in 73ae298f0aa4a5e1ea45603f8d56ece3e762a70d

Mark-H commented:
I goofed up merging this so had to revert and apply it again - sorry for the confusion.

smg6511 commented:
> Thank you for your hard work on this @smg6511 - took a while to get it done, but we got there, eventually. Appreciate your _persistence_!

Thanks for getting this in at last! There's more I'd been working on waiting in the wings in terms of grids improvements (but not nearly of this magnitude) ;-)
